### PR TITLE
Fixes #9333: If we change the global policy mode, promises are not generated if we click on "update policies"

### PR DESCRIPTION
**Pull request by fanf**
https://www.rudder-project.org/redmine/issues/9333

### CODE REVIEW

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala [hunk 1]
```diff
@@ -52,86 +52,137 @@ import net.liftweb.common.Loggable
 import com.normation.cfclerk.domain.Variable
 import com.normation.rudder.services.policies.write.Cf3PolicyDraftId
 import com.normation.rudder.services.policies.write.Cf3PolicyDraft
+import com.normation.rudder.domain.reports.NodeModeConfig
 
 
-case class PolicyCache(
+case class policyHash(
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala]: why the lowercase for Policy ?

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala [hunk 2]
```diff
@@ -52,86 +52,137 @@ import net.liftweb.common.Loggable
...
+     * A parameter update must lead to a regeneration of the node,
+     * because it could be used in pur CFEngine (when used in
+     * directive, we already tracked it with directives values).
+     *
+     * So a modification in ANY parameter lead to a regeneration
+     * of ALL NODES (everything, yes).
      *
-     * Also, variable with no values are equivalent to no variable, so we
-     * remove them.
+     * For parameters, do not consider:
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala]: don't you mean properties ? as parameters are overridable, but there is nothing to override them

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala [hunk 3]
```diff
@@ -52,86 +52,137 @@ import net.liftweb.common.Loggable
...
-     * for node info, only consider:
-     * name, hostname, agentsName, policyServerId, localAdministratorAccountName
+     * Take into account anything that has influence on policies but
+     * without being only a directive parameter (that latter case will
+     * be handle directly at the directive level) or a system variable
+     * (which is already in nodeContext)
+     *
+     * - all ${rudder.node} params (because can be used in pure CFEngine)
+     * - node properties (because can be used in pure CFEngine)
+     * - isPolicyServer (for nodes becoming relay)
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala]: <3

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala [hunk 4]
```diff
@@ -52,86 +52,137 @@ import net.liftweb.common.Loggable
...
+     * - serverRoles (because not the same set of directives - but
+     *   perhaps it is already handle in the directives)
+     * - agentsName (because directly used in templates for if conditions)
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala]: <3

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala [hunk 5]
```diff
@@ -142,42 +193,49 @@ object NodeConfigurationCache {
...
         id = nodeConfig.nodeInfo.id
       , writtenDate = nodeConfig.writtenDate
-      , nodeInfoCache = nodeInfoCacheValue
-      , parameterCache = parameterCache
-      , nodeContextCache = nodeContextCache
-      , policyCache = policyCacheValue
+      , nodeInfoHash = nodeInfoHashValue
+      , parameterHash = parameterHash
+      , nodeContextHash = nodeContextHash
+      , policyHash = policyHashValue
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala]: what about the properties, shouldn't you have a cache of it ?

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationServiceImpl.scala [hunk 6]
```diff
@@ -369,17 +193,23 @@ class NodeConfigurationServiceImpl(
...
+  override def detectChangeInNodes(nodes : Seq[NodeConfiguration], cache: Map[NodeId, NodeConfigurationHash], directiveLib: FullActiveTechniqueCategory) : Set[RuleId] = {
+
+    // changes in nodes don't imply changes in directives / rules / serial.
+    // these change will lead to a new config id and generation if needed.
+    // we only increment serial when the cache is empty, which mean that it
+    // is a new rudder, or a "clear cache".
+    // Note that if the cache is empty, all nodes will be rewriten in all
+    // cases, but that allows to check that nodes really get their
+    // policies (because node config id may be the same, but they will continue sending
+    // reports with the previous serial)
```
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationServiceImpl.scala]: given the comment and the code, the name of the method is super-confusing
Could it be "initializeNodeCache" or something like that ?
Plus ,the directiveLib isn't used at all, you can remove it from the signature

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala (hunk 1 above)
Comment (fanf) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala]: ah. Automatic renamaed didn't do what it was supposed to do.

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala (hunk 5 above)
Comment (fanf) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationCacheRepository.scala]: It is part in nodeInfoHash. Why a new hash for them ?

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

File: rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationServiceImpl.scala (hunk 6 above)
Comment (ncharles) [inline rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationServiceImpl.scala]: thank you

**Comment by ncharles** [comment]
that's good for me ! thank you

**Pull request merged**